### PR TITLE
[4.4] Rebuild mod_footer with new Joomla 4 structure

### PR DESCRIPTION
**Pull request by sandewt**
Pull Request for Issue # .

### Summary of Changes

Rebuild mod_footer with new Joomla 4 structure
See: testing Instructions

### Testing Instructions

Just try to use it as usual.

### Actual result BEFORE applying this Pull Request

Module works

### Expected result AFTER applying this Pull Request

Module works

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

Related to #39931 (mod_custom). Thks @sergeytolkachyov and @joomdonation 

[EDIT mod_footer system test #40114]

### CODE REVIEW

File: modules/mod_footer/src/Dispatcher/Dispatcher.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+     *
+     * @return  array
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected function getLayoutData(): array
+    {
+        $data = parent::getLayoutData();
+
+        $data['list'] = $this->getHelperFactory()->getHelper('FooterHelper')->getList($this->getApplication());
```
Comment (joomdonation) [inline modules/mod_footer/src/Dispatcher/Dispatcher.php]: I think for simple string like that, you do not need to create a Helper to get that data. The code here could just be changed to:

```php
$lineOne = Text::_('MOD_FOOTER_LINE1');
$lineOne = str_replace('%date%', HTMLHelper::_('date', HTMLHelper::_('date', 'Now', 'Y'), 'Y'), $lineOne);
$lineOne = str_replace('%sitename%', $this->getApplication()->get('sitename'), $lineOne);

$data['lineone'] = $lineOne;
```

File: modules/mod_footer/services/provider.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+     * @param   Container  $container  The DI container.
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function register(Container $container)
+    {
+        $container->registerServiceProvider(new ModuleDispatcherFactory('\\Joomla\\Module\\Footer'));
+        $container->registerServiceProvider(new HelperFactory('\\Joomla\\Module\\Footer\\Site\\Helper'));
```
Comment (joomdonation) [inline modules/mod_footer/services/provider.php]: This line could be removed, see my comment below (no need to use Helper for simple string data in this module)

File: modules/mod_footer/src/Helper/FooterHelper.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * @package     Joomla.Site
+ * @subpackage  mod_footer
+ *
+ * @copyright   (C) 2005 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+namespace Joomla\Module\Footer\Site\Helper;
```
Comment (joomdonation) [inline modules/mod_footer/src/Helper/FooterHelper.php]: This file could be removed.

File: modules/mod_footer/tmpl/default.php [hunk 4]
```diff
@@ -14,6 +14,6 @@
 
 ?>
 <div class="mod-footer">
-    <div class="footer1"><?php echo $lineone; ?></div>
+    <div class="footer1"><?php echo $list; ?></div>
```
Comment (joomdonation) [inline modules/mod_footer/tmpl/default.php]: Please revert the change here. We need to keep using original variable name $lineone in `getLayoutData` method of the dispatcher, otherwise, if someone overrides this module layout in the template, the module will be broken with this update.

**Review by joomdonation — COMMENTED** [review]
Thanks for working on on this @sandewt . Overall, it looks good. Just some minor suggestions from me and I think it should be fine.

**Review by sergeytolkachyov — COMMENTED** [review]
I would be careful when changing the layout files, since many people may have redefinitions that are in the tmpl module or in the template and their files will have old variable names, which will cause an error when updating Joomla and force them to look for the cause.

File: modules/mod_footer/src/Dispatcher/Dispatcher.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+// phpcs:enable PSR1.Files.SideEffects
+
+/**
+ * Dispatcher class for mod_footer
+ *
+ * @since  __DEPLOY_VERSION__
+ */
+class Dispatcher extends AbstractModuleDispatcher
+{
+    //use HelperFactoryAwareTrait;
```
Comment (joomdonation) [inline modules/mod_footer/src/Dispatcher/Dispatcher.php]: Please remove this commented code.

File: modules/mod_footer/src/Dispatcher/Dispatcher.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+     */
+    protected function getLayoutData(): array
+    {
+        $data = parent::getLayoutData();
+
+        $lineOne = Text::_('MOD_FOOTER_LINE1');
+        $lineOne = str_replace('%date%', HTMLHelper::_('date', HTMLHelper::_('date', 'Now', 'Y'), 'Y'), $lineOne);
+        $lineOne = str_replace('%sitename%', $this->getApplication()->get('sitename'), $lineOne);
+
+        $data['lineOne'] = $lineOne;
```
Comment (joomdonation) [inline modules/mod_footer/src/Dispatcher/Dispatcher.php]: Please change it to the below code. As mentioned, you cannot change the variable name here because it will break the module if someone overrides the layout before

```php
$data['lineone'] = $lineOne;
```

File: modules/mod_footer/tmpl/default.php [hunk 7]
```diff
@@ -14,6 +14,6 @@
 
 ?>
 <div class="mod-footer">
-    <div class="footer1"><?php echo $lineone; ?></div>
+    <div class="footer1"><?php echo $lineOne; ?></div>
```
Comment (joomdonation) [inline modules/mod_footer/tmpl/default.php]: Please revert this change. As mentioned earlier, we could not change the variable name passed to module layout due to backward compatible promise.

**Review by joomdonation — COMMENTED** [review]
Thanks. There are still some minor things left and we should be good.

File: modules/mod_footer/src/Dispatcher/Dispatcher.php (hunk 6 above)
Comment (sandewt) [reply modules/mod_footer/src/Dispatcher/Dispatcher.php]: Done

**Comment by laoneo** [comment]
Can you guys rebase this and the other pr's which do convert the modules to service providers to the 4.4 branch?

File: modules/mod_footer/src/Dispatcher/Dispatcher.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+     * @return  array
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected function getLayoutData(): array
+    {
+        $data = parent::getLayoutData();
+
+        $lineOne = Text::_('MOD_FOOTER_LINE1');
+        $lineOne = str_replace('%date%', HTMLHelper::_('date', HTMLHelper::_('date', 'Now', 'Y'), 'Y'), $lineOne);
```
Comment (joomdonation) [inline modules/mod_footer/src/Dispatcher/Dispatcher.php]: There is a small typo here. Please change the code to:

```php
$lineOne = str_replace('%date%', HTMLHelper::_('date', 'now', 'Y'), $lineOne);
```

Comment (sandewt) [reply modules/mod_footer/src/Dispatcher/Dispatcher.php]: Thanks

**Comment by sandewt** [comment]
> Can you guys rebase this and the other pr's which do convert the modules to service providers to the 4.4 branch?

How can I tconvert this PR in an easy way?

**Comment by laoneo** [comment]
Click on the edit button next to the title and then change the target branch.

**Comment by sandewt** [comment]
> Click on the edit button next to the title and then change the target branch.

Done.

Does the `Build: 4.3-dev` label change automatically in the Issue Tracker CMS above? Or can I change that in `Build: 4.4-dev`?
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40066">issues.joomla.org/tracker/joomla-cms/40066</a>.</sub>

**Comment by laoneo** [comment]
@Quy changed it. If you want to join the bug squad team, then you get the permissions to do that by your own.

File: modules/mod_footer/src/Dispatcher/Dispatcher.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * @package     Joomla.Site
+ * @subpackage  mod_footer
+ *
+ * @copyright   (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+namespace Joomla\Module\Footer\Site\Dispatcher;
```
Comment (sandewt) [inline modules/mod_footer/src/Dispatcher/Dispatcher.php:11]: @joomdonation 

The `FooterHelper.php` file is deleted.

Can this line be removed?  `namespace Joomla\Module\Footer\Site\Dispatcher;` 
Or is there possibly a reason not to do this?

Comment (joomdonation) [reply modules/mod_footer/src/Dispatcher/Dispatcher.php:11]: @sandewt It needs to be there, please don't remove it. If you remove it, the class name will just be `Dispatcher`, not `\Joomla\Module\Footer\Site\Dispatcher\Dispatcher` and `ModuleDispatcherFactory` could not find the class. Further more, there could be multiple files with same class name `Dispatcher` and causes fatal error

Sorry, I don't know how to explain more clear but it needs to be kept, don't remove it.

Comment (sandewt) [reply modules/mod_footer/src/Dispatcher/Dispatcher.php:11]: The strange thing is that when I test locally I can just remove this line with no effect. The footer remains visible.

Comment (sandewt) [reply modules/mod_footer/src/Dispatcher/Dispatcher.php:11]: Thanks anyway. The line just stands.

Comment (joomdonation) [reply modules/mod_footer/src/Dispatcher/Dispatcher.php:11]: I can only guess that you haven't deleted the `modules/mod_footer/mod_footer.php` in your installation yet. If you delete that file, I'm sure footer won't be displayed anymore

Comment (sandewt) [reply modules/mod_footer/src/Dispatcher/Dispatcher.php:11]: > I can only guess that you haven't deleted the modules/mod_footer/mod_footer.php

Indeed, you saw that right. 👍

**Comment by joomdonation** [comment]
I have tested this item :white_check_mark: successfully on 6a13879a76bb89670d4330be7b68d1b06a5f59ae<br /><br />Thanks for your work !<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40066">issues.joomla.org/tracker/joomla-cms/40066</a>.</sub>

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on 6a13879a76bb89670d4330be7b68d1b06a5f59ae<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40066">issues.joomla.org/tracker/joomla-cms/40066</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40066">issues.joomla.org/tracker/joomla-cms/40066</a>.</sub>

**Comment by richard67** [comment]
@laoneo Site note: For the deleted files list in script.php I will care like I do for the other branches.

File: modules/mod_footer/src/Dispatcher/Dispatcher.php [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+     *
+     * @return  array
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected function getLayoutData(): array
+    {
+        $data = parent::getLayoutData();
+
+        $lineOne = Text::_('MOD_FOOTER_LINE1');
```
Comment (laoneo) [inline modules/mod_footer/src/Dispatcher/Dispatcher.php]:
```suggestion
        $lineOne = $this->getApplication()->getLanguage()->_('MOD_FOOTER_LINE1');
```

File: modules/mod_footer/src/Dispatcher/Dispatcher.php [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        $lineOne = str_replace('%date%', HTMLHelper::_('date', 'now', 'Y'), $lineOne);
+        $lineOne = str_replace('%sitename%', $this->getApplication()->get('sitename'), $lineOne);
```
Comment (laoneo) [inline modules/mod_footer/src/Dispatcher/Dispatcher.php]:
```suggestion
        $lineOne = str_replace('%sitename%', $this->getApplication()->get('sitename', ''), $lineOne);
```

File: modules/mod_footer/src/Dispatcher/Dispatcher.php (hunk 10 above)
Comment (sandewt) [reply modules/mod_footer/src/Dispatcher/Dispatcher.php]: Why not `Text::_('MOD_FOOTER_LINE1')` ?

`Text` is not deprecated ? I'm curious about your argument.

Comment (laoneo) [reply modules/mod_footer/src/Dispatcher/Dispatcher.php]: It's better not to use static classes as it gets harder to write unit tests for it.

**Comment by laoneo** [comment]
Added a system test for the footer module in #40114. Now we see if the test succeeds here as well.

**Comment by laoneo** [comment]
Thanks!
